R/register_handlers.R

```diff
@@ -0,0 +1,98 @@
+#' Register handlers for logging messages, warnings and errors
+#'
+#' @param namespace (`character(1)`) the logger namespace
+#' @param package (`character(1)`) the package name
+#'
+#' @return `NULL` invisibly. Called for its side effects.
+#'
+#' @details This function registers global handlers for messages, warnings and errors.
+#' The handlers will investigate the call stack and if it contains a function
+#' from the package, the message, warning or error will be logged into the respective
+#' namespace.
+#' The handlers are registered only once per package and type. Consecutive calls will be ignored.
+#'
+#' @note Registering handlers is disallowed within `tryCatch` blocks which is used throughout
+#' base R and other packages. Because of this, the `register_handlers` function is wrapped in a
+#' `try` block to avoid errors.
+#'
+#' @seealso [globalCallingHandlers()]
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' register_logger(namespace = "teal.logger")
+#' register_handlers("teal.logger")
+#' # see the outcome
+#' globalCallingHandlers()
+#' }
+register_handlers <- function(namespace, package = namespace) {
+  try(
+    {
+      register_handler_type(namespace = namespace, package = package, type = "message")
+      register_handler_type(namespace = namespace, package = package, type = "warning")
+      register_handler_type(namespace = namespace, package = package, type = "error")
+    },
+    silent = TRUE
+  )
+
+  invisible(NULL)
+}
+
+#' @keywords internal
+register_handler_type <- function(
+    namespace,
+    package,
+    type = c("error", "warning", "message")) {
+  match.arg(type)
+  if (!(is.character(namespace) && length(namespace) == 1)) {
+    stop("namespace argument must be a scalar character.")
+  }
+  if (!(namespace %in% logger::log_namespaces())) {
+    stop("namespace argument must be a pre-registered logger namespace.")
+  }
+  if (!(is.character(package) && length(package) == 1)) {
+    stop("package argument must be a scalar character.")
+  }
+
+  logger_fun <- switch(type,
+    error = logger::log_error,
+    warning = logger::log_warn,
+    message = logger::log_info
+  )
+
+  # avoid re-registering the same handler
+  gch <- globalCallingHandlers()[names(globalCallingHandlers()) == type]
+  if (any(sapply(gch, attr, "type") == "teal.logger_handler" & sapply(gch, attr, "package") == package)) {
```

Comment:
what if `gch` is `list()` (empty list)? should this condition be avoided, or it fails because `any()` gives `FALSE` on `logical(0)` that comes from 

```r
> sapply(list(), attr, "type") == "teal.logger_handler"
logical(0)
```

Reply:
maybe wrap operation on `gch` under `if(length(globalCallingHandlers()))` statement

Reply:
This step suppose to avoid re-registering the same object. I need to look-up in the list, check if a given object is there - if yes then leave earlier.
If a list is empty then this means that no objects had been registered before. I can safely proceed to register action and there is no need to leave earlier.
On that note I think that the current behaviour is correct. For clarity, I might introduce a new condition check if a list is not (!) empty and then move on to check its elements.